clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java

```diff
@@ -292,29 +415,66 @@ public static double readDouble(ByteBuffer buffer) {
      * @param buffer The output to write to
      */
     public static void writeUnsignedVarint(int value, ByteBuffer buffer) {
```

Comment:
What does netty do for this? Similar to what we're doing here or something else?

Reply:
Netty uses the default loop based implementation: https://github.com/netty/netty/blob/5d1f99655918c9c034ca090d51b64eced73f742f/codec/src/main/java/io/netty/handler/codec/protobuf/ProtobufVarint32LengthFieldPrepender.java#L58 

I wasn't able to find a project which does what we are doing [except for the blog post](https://steinborn.me/posts/performance/how-fast-can-you-write-a-varint/) I mentioned above and over [here](https://github.com/richardstartin/varints/blob/main/src/main/java/io/github/richardstartin/varints/SmartNoDataDependencyVarIntState.java).

Note that protobuf uses unrolled implementation for it's c++ code at 
https://github.com/protocolbuffers/protobuf/blob/2dc5338ea222e1f4e0357e46b702ed6a0e82aaeb/src/google/protobuf/io/coded_stream.h#L913 Not that it matters for us since compilers are different from Java & c++ but adding it here as a data point.